File: customview/src/main/java/com/bksx/customview/customview/SportView.java
package com.bksx.customview.customview;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import com.bksx.customview.Utils;

import androidx.annotation.Nullable;

public class SportView extends View {

    Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private static final float RADIUS = Utils.dp2px(150);
    private static final float RING_WIDTH = Utils.dp2px(20);
    private static final float PROGRESS_WIDTH = Utils.dp2px(18);
    Rect bounds = new Rect();
    Paint.FontMetrics metrics = new Paint.FontMetrics();
    public SportView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    {
        paint.setTextSize(Utils.dp2px(100));
        paint.setTypeface(Typeface.MONOSPACE);
        paint.setTextAlign(Paint.Align.CENTER);
    }
    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        //画环
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeWidth(RING_WIDTH);
        paint.setColor(Color.parseColor("#C4C4C4"));
        canvas.drawCircle(getWidth() / 2,getHeight() / 2 , RADIUS,paint);

        //画进度
        paint.setStrokeWidth(PROGRESS_WIDTH);
        paint.setColor(Color.parseColor("#CB0F83"));
        paint.setStrokeCap(Paint.Cap.ROUND);
        canvas.drawArc(getWidth() / 2 - RADIUS,getHeight()  / 2 - RADIUS, getWidth() / 2 +RADIUS,
                getHeight() / 2 + RADIUS,-90,225,false,paint);

        //画文字
        String text = "sury";
        paint.setStyle(Paint.Style.FILL);
//        paint.measureText(text);
//        paint.getTextBounds(text,0,text.length(),bounds);//适合静态文字居中
        paint.getFontMetrics(metrics);
        float offset = (metrics.descent + metrics.ascent) / 2f;//文字高度中心到当前基线（初始认为的正确显示位置）的距离，同时也是基线需要偏移的距离
        canvas.drawText(text,getWidth() / 2 ,getHeight()  / 2 - offset,paint);

        //画文字贴边
        paint.setTextAlign(Paint.Align.LEFT);
        paint.setTextSize(Utils.dp2px(150));
        paint.getTextBounds(text,0,text.length(),bounds);
        float small = bounds.top;
        canvas.drawText(text,-bounds.left,-bounds.top,paint);

        //另外一行小字
        paint.setTextSize(Utils.dp2px(15));
        paint.getTextBounds(text,0,text.length(),bounds);
        canvas.drawText(text,-bounds.left,-small + paint.getFontSpacing(),paint);
    }
}

File: hencoderplus/src/main/java/com/example/hencoderplus/customview/SportView.kt
package com.example.hencoderplus.customview

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.Log
import android.view.View
import kotlin.math.abs


class SportView(context: Context?, attrs: AttributeSet?) : View(context, attrs) {
    private var paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var bounds = Rect()
    private var metrics = Paint.FontMetrics()
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        //画环
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = RING_WIDTH
        paint.color = Color.parseColor("#C4C4C4")
        canvas.drawCircle(width / 2.toFloat(), height / 2.toFloat(), RADIUS, paint)
        //画进度
        paint.strokeWidth = PROGRESS_WIDTH
        paint.color = Color.parseColor("#CB0F83")
        paint.strokeCap = Paint.Cap.ROUND
        canvas.drawArc(width / 2 - RADIUS, height / 2 - RADIUS, width / 2 + RADIUS,
                height / 2 + RADIUS, -90f, 225f, false, paint)
        //画文字
        val text = "sury"
        paint.style = Paint.Style.FILL
        //两种垂直居中方式:bounds,metrics即静态和动态
//        paint.getTextBounds(text,0,text.length,bounds)
//        val offset = (bounds.top + bounds.bottom) / 2
//        val top = bounds.top
//        val bottom = bounds.bottom
//        Log.d("ccg","top:$top,bottom:$bottom")
        // 文字baseline在y轴方向的位置
        paint.getFontMetrics(metrics)
        val offset = (metrics.ascent + metrics.descent) / 2
        canvas.drawText(text, (width/2).toFloat(), (height/2).toFloat()-offset,paint)

        //文字贴边
        paint.textAlign = Paint.Align.LEFT
        paint.textSize = dp2px(70f)
        paint.getTextBounds(text, 0, text.length, bounds)
        canvas.drawText(text, (-bounds.left).toFloat(), -bounds.top.toFloat(), paint)


    }

    companion object {
        private val RADIUS = dp2px(150f)
        private val RING_WIDTH = dp2px(20f)
        private val PROGRESS_WIDTH = dp2px(18f)
    }

    init {
        paint.textSize = dp2px(100f)
        paint.typeface = Typeface.MONOSPACE
        paint.textAlign = Paint.Align.CENTER
    }
}